src/System.Linq/src/System/Linq/Select.SpeedOpt.cs

```diff
@@ -157,6 +157,127 @@ public TResult TryGetLast(out bool found)
             }
         }
 
+        private sealed partial class SelectRangeIterator<TResult> : Iterator<TResult>, IPartition<TResult>
+        {
+            private readonly int _start;
+            private readonly int _end;
+            private readonly Func<int, TResult> _selector;
+
+            public SelectRangeIterator(int start, int end, Func<int, TResult> selector)
+            {
+                Debug.Assert(start < end);
+                Debug.Assert((end - start) <= int.MaxValue);
+                Debug.Assert(selector != null);
+
+                _start = start;
+                _end = end;
+                _selector = selector;
+            }
+
+            public override Iterator<TResult> Clone() =>
+                new SelectRangeIterator<TResult>(_start, _end, _selector);
+
+            public override bool MoveNext()
+            {
+                if (_state < 1 || _state == (_end - _start + 1))
+                {
+                    Dispose();
+                    return false;
+                }
+
+                int index = _state++ - 1;
+                _current = _selector(_start + index);
+                return true;
+            }
+
+            public override IEnumerable<TResult2> Select<TResult2>(Func<TResult, TResult2> selector) =>
+                new SelectRangeIterator<TResult2>(_start, _end, CombineSelectors(_selector, selector));
+
+            public TResult[] ToArray()
+            {
+                var results = new TResult[_end - _start];
+                int srcIndex = _start;
+                for (int i = 0; i < results.Length; i++)
+                {
+                    results[i] = _selector(srcIndex++);
+                }
+
+                return results;
+            }
+
+            public List<TResult> ToList()
+            {
+                var results = new List<TResult>(_end - _start);
+                for (int i = _start; i != _end; i++)
+                {
+                    results.Add(_selector(i));
+                }
+
+                return results;
+            }
+
+            public int GetCount(bool onlyIfCheap)
+            {
+                // In case someone uses Count() to force evaluation of the selector,
+                // run it provided `onlyIfCheap` is false.
+                if (!onlyIfCheap)
+                {
+                    for (int i = _start; i != _end; i++)
+                    {
+                        _selector(i);
+                    }
+                }
+
+                return _end - _start;
+            }
+
+            public IPartition<TResult> Skip(int count)
+            {
+                if (count >= (_end - _start))
+                {
+                    return EmptyPartition<TResult>.Instance;
+                }
+
+                return new SelectRangeIterator<TResult>(_start + count, _end, _selector);
+            }
+
+            public IPartition<TResult> Take(int count)
```

Comment:
Can `count` be < 0?

Reply:
> Can count be < 0?

No; the call sites check for 0 and return Empty in that case.  I can add an assert.